### PR TITLE
Adding optional fallback to encode java.sql.Date as EpochMs

### PR DESCRIPTION
**Pull request by FabianMeiswinkel**
# Description
The Cosmos DB Spark connector in Spark 2.4 used to convert both `java.sql.Date` and `java.sql.Timestamp` into Epoch Milliseconds. IN Spark 3 this conversion has changed because Spark 3 has a catalyst optimizer that would convert `java.sql.Date` into an Int of EpochDays and `java.sql.Timestamp` into a Long of Epoch Microseconds.
This PR add a configuration override to allow falling back to the previous conversion (`java.sql.Date` and `java.sql.Timestamp` will be converted into Epoch Milliseconds)

# All SDK Contribution checklist:
- [x] **The pull request does not introduce [breaking changes]**
- [x] **CHANGELOG is updated for new features, bug fixes or other significant changes.**
- [x] **I have read the [contribution guidelines](https://github.com/Azure/azure-sdk-for-java/blob/main/CONTRIBUTING.md).**

## [General Guidelines and Best Practices](https://github.com/Azure/azure-sdk-for-java/blob/main/CONTRIBUTING.md#developer-guide)
- [x] Title of the pull request is clear and informative.
- [x] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/cleaning-up-commits.md).

### [Testing Guidelines](https://github.com/Azure/azure-sdk-for-java/blob/main/CONTRIBUTING.md#building-and-unit-testing)
- [x] Pull request includes test coverage for the included changes.

### CODE REVIEW

**Review by ealsur — APPROVED** [review]

File: sdk/cosmos/azure-cosmos-spark_3_2-12/src/main/scala/com/azure/cosmos/spark/CosmosRowConverter.scala [hunk 1]
```diff
@@ -293,14 +295,40 @@ private[cosmos] class CosmosRowConverter(
...
+            case SerializationDateTimeConversionModes.AlwaysEpochMilliseconds =>
+              convertToJsonNodeConditionally(rowData.asInstanceOf[Long].asInstanceOf[Instant].toEpochMilli)
+          }
         case TimestampType if rowData.isInstanceOf[java.lang.Integer] =>
-          convertToJsonNodeConditionally(rowData.asInstanceOf[Integer])
+          serializationConfig.serializationDateTimeConversionMode match {
+            case SerializationDateTimeConversionModes.Default =>
+              convertToJsonNodeConditionally(rowData.asInstanceOf[java.lang.Integer])
+            case SerializationDateTimeConversionModes.AlwaysEpochMilliseconds =>
+              convertToJsonNodeConditionally(rowData.asInstanceOf[java.lang.Integer].asInstanceOf[Instant].toEpochMilli)
```
Comment (xinlian12) [inline sdk/cosmos/azure-cosmos-spark_3_2-12/src/main/scala/com/azure/cosmos/spark/CosmosRowConverter.scala:330]: can we cast from Integer to Instant directly?

File: sdk/cosmos/azure-cosmos-spark_3_2-12/src/test/scala/com/azure/cosmos/spark/CosmosRowConverterSpec.scala [hunk 2]
```diff
@@ -423,6 +456,42 @@ class CosmosRowConverterSpec extends UnitSpec with BasicLoggingTrait {
     objectNode.get(colName4).asInt() shouldEqual colVal3
   }
 
+  "date and time in spark row" should "should honor dateTimeConversionMode config" in {
```
Comment (xinlian12) [inline sdk/cosmos/azure-cosmos-spark_3_2-12/src/test/scala/com/azure/cosmos/spark/CosmosRowConverterSpec.scala:459]: This test only test the `convertSparkDataTypeToJsonNodeNonNull`
can we also add a test for `convertSparkDataTypeToJsonNodeConditionally`

File: sdk/cosmos/azure-cosmos-spark_3_2-12/src/test/scala/com/azure/cosmos/spark/CosmosRowConverterSpec.scala [hunk 3]
```diff
@@ -423,6 +456,42 @@ class CosmosRowConverterSpec extends UnitSpec with BasicLoggingTrait {
...
+    convertToCatalyst(colVal1Raw).isInstanceOf[Int] shouldEqual true
+    val colVal1= convertToCatalyst(colVal1Raw).asInstanceOf[Int]
+    colVal1 shouldEqual -8786
+    colVal1 shouldEqual testDate.toEpochDay
+
+    // Catalyst optimizer will convert java.sql.Timestamp into epoch Microseconds
+    val colVal2Raw = Timestamp.from(testTimestamp)
+    convertToCatalyst(colVal2Raw).isInstanceOf[Long] shouldEqual true
+    val colVal2= convertToCatalyst(colVal2Raw).asInstanceOf[Long]
+    colVal2 shouldEqual -727530468000000L
```
Comment (xinlian12) [inline sdk/cosmos/azure-cosmos-spark_3_2-12/src/test/scala/com/azure/cosmos/spark/CosmosRowConverterSpec.scala:478]: hmm, it failed for me:
```
-727501668000000 did not equal -727530468000000
ScalaTestFailureLocation: com.azure.cosmos.spark.CosmosRowConverterSpec at (CosmosRowConverterSpec.scala:478)
Expected :-727530468000000
Actual   :-727501668000000
```

File: sdk/cosmos/azure-cosmos-spark_3_2-12/src/main/scala/com/azure/cosmos/spark/CosmosRowConverter.scala (hunk 1 above)
Comment (FabianMeiswinkel) [reply sdk/cosmos/azure-cosmos-spark_3_2-12/src/main/scala/com/azure/cosmos/spark/CosmosRowConverter.scala:330]: No - but this conversion would have thrown ClassCastException anyway - missed to port the fix from convertSparkDataTypeToJosnNodeNonNull

**Review by xinlian12 — APPROVED** [review]
LGTM, thanks

**Pull request merged**